Src/Particle/AMReX_NeighborParticlesGPUImpl.H

```diff
@@ -185,28 +143,39 @@ buildNeighborCopyOp ()
 
         auto& src_tile = plev[index];
         auto& aos   = src_tile.GetArrayOfStructs();
-        const size_t np = aos.numParticles();
+        const size_t np_real = aos.numParticles();
 
-        Array4<const int> const& mask_arr = m_neighbor_mask_ptr->array(mfi);
+        size_t np = np_real;
+        if (use_boundary_neighbor) {
+            np = m_boundary_particle_ids[lev][index].size();
+        }
+        else {
+            m_boundary_particle_ids.resize(1);
+            m_boundary_particle_ids[lev][index];
+        }
+
+        const auto*  p_bndry_pid = m_boundary_particle_ids[lev][index].dataPtr();
 
         Gpu::DeviceVector<int> counts(np + 1, 0);
         Gpu::DeviceVector<int> offsets(np + 1);
         auto p_counts = counts.dataPtr();
         auto p_offsets = offsets.dataPtr();
 
         ParticleType* p_ptr = &(aos[0]);
-        auto p_code_array = m_code_array[gid].dataPtr();
-        auto p_code_offsets = m_code_offsets[gid].dataPtr();
+        auto p_code_array   = m_code_array[gid].dataPtr();
+        auto p_isec_boxes   = m_isec_boxes[gid].dataPtr();
+        const int nisec_box = m_isec_boxes[gid].size();
+        // auto p_code_offsets = m_code_offsets[gid].dataPtr();
+
         AMREX_FOR_1D ( np, i,
         {
-            IntVect iv = getParticleCell(p_ptr[i], plo, dxi, domain);
-            int code = mask_arr(iv);
-            if (code >= 0)
-            {
-                int start = p_code_offsets[code];
-                int stop =  p_code_offsets[code+1];
-                p_counts[i] += stop - start;
-            }
+            // note that cannot use ternary statement here because p_bndry is not
+            // properly allocated when use_boundary_neighbor=false
+            int pid = i;
+            if (use_boundary_neighbor)  pid = p_bndry_pid[i];
```

Comment:
Minor style nit: could you use curly braces around this `if` statement, and the `for` loop below, even though they are only one line, as per [here](https://github.com/AMReX-Codes/amrex/blob/development/CONTRIBUTING.md#amrex-coding-style-guide)?